### ISSUE TITLE
`$` broken for platforms where int < int32

### ISSUE THREAD

**Issue by PMunch** [nim-lang/Nim#20828]
### What happened?

I ran into this when implementing the BME280 sensor for Ratel. During debugging I wanted to print out some variables, amongst those an `int32`. When being converted to a string with `$` and printed out over serial the value shown was a medium size negative number, while I expected a 3x larger positive number. Digging into the C code I realized that before being passed over to `$` the number was being cast to an int. Looking at the `system/dollars` module it was obvious that this happens because there is no `$` defined for `int32`, rather it depends on `int` being the largest of the two and converts `int32` to an `int`. This is a problem on a platform such as the one I was targeting in my test which uses 16 bit integers, meaning that `int` is actually `int16`. The `int32` is converted to an `int16` by simply discarding the top two bytes, which meant that if my number happens to have the highest 16th bit set it gets treated as a negative number.

### Nim Version

Tested on 1.6.8, 1.6.10rc, and devel

### Possible Solution

Add more explicit `$` procs or create these on compile-time based on the biggest integer for the platform.

### Additional Information
With the newly published article pushing Nim for embedded platforms these kinds of bugs should be a high priority to fix.

**Comment by Araq** [comment]
> With the newly published article pushing Nim for embedded platforms these kinds of bugs should be a high priority to fix.

Maybe but embedded CPUs can be 32 bits too.

**Comment by PMunch** [comment]
Fair enough, although quite a few of them are 16-bit.. Turns out this isn't just an issue for strings though. I had a check like this:
```nim
var1 = if var1 < 0: 0 else: var1
```
in my code which seems all fine. The check works as expected since `var1 < 0` does a comparison between two int32's. But the `0 else: var1` part causes trouble. That first `0` get treated as an int, which means the whole block returns an int, which Nim thinks is fine even though `var1` is defined as a int32 because it thinks it can fit an int32 in an int. So that check trims two bytes of `var1` whenever it is larger than 0.. This bug took my hours to figure out over multiple days by the way.

**Comment by Araq** [comment]
These are great points for an RFC "why mapping int to int16 for embedded does not work and what to do about it".

**Comment by bung87** [comment]
after breaking some codes, I think this one is not fit the label as "easy"

**Comment by PMunch** [comment]
> These are great points for an RFC "why mapping int to int16 for embedded does not work and what to do about it".

That's a tough choice though. `int` in Nim is defined to be the size of a pointer, and pointers on this architecture are 16-bit. I imagine the same issue appears on 32-bit architectures when using `int64`.

**Comment by Araq** [comment]
You can make the case that `int` should be mapped to int32 regardless because the code in the wild is not tested on 16 bit platforms. But I prefer to map int to int16 and deal with the fallout. :-)

**Comment by PMunch** [comment]
Yeah I also prefer to have int be a target dependent thing which makes sense. I tried to trigger this bug on 32 bit machines by the way. But the statement: `num = if num < 0: 0 else: num` fails when `num` is int64, even on 64-bit platforms. So there is some logic in the compiler to try and prevent this kind of bug. But it doesn't do anything when `num` is 32 bits or lower, it just somehow deals with it.
